### PR TITLE
Framegraph: record renderpasses and their dependencies

### PR DESCRIPTION
**Pull request by hevrard**
Add the structure to go over subcommands, record renderpasses, and
establish their dependencies based on the dependency graph. For now,
renderpasses have very little information attached to them: this is on
purpose, to keep this commit small.

The main part of this commit is the inter-renderpass dependencies
logic, in updateDependencies().

The change in MutateWithSubcommands() is to allow passing a nil
postCmdCb to it.

Bug: b/172218304
Test: manually, on a few traces.

### CODE REVIEW

**Comment by hevrard** [comment]
For illustration, this steps let us get this framegraph on Sascha Willems "deferred" demo:

framegraph

**Review by johnplate — APPROVED** [review]
LGTM, but wait for another approval

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:55]: Do we know that i has a type of `CommandReferenceʳ` ? If yes, we should replace it with `interface{}`? if no, then we should check if type conversion is valid. Otherwise it's a direct crash.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:53]: I can see that this is a form of closure that modifies the function fields several times in a loop in `MutateWithSubcommands` but this is quite subtle and can be misleading. 

This might be quite common in functional paradigm which I am not familiar with but as the rest of the code is quite non-functional(no puns intended) I think having a small helper struct with one function that actually does the operation and passing a lambda that calls the particular function in the object would be much cleaner.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:113]: A notifier: This is a map and in Golang this will iterate differently for every range operation for the same data. If this is considered there is no problem.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:137]: Future comment: This does not do anything for now but if you plan to make type assertion on the `struct{}` that's a lot of type assertion there.

File: gapis/api/vulkan/framegraph.go
Comment (yalcinmelihyasin) [inline gapis/api/vulkan/framegraph.go:145]: I don't know how does go compiler handles this(maybe realises and optimizes) but instead of creating a new lambda function for every rpInfo in rpInfos this might be converted into a normal recursive function calls. But not the highest priority I would say especially if it will take long time.

**Review by yalcinmelihyasin — CHANGES_REQUESTED** [review]
I think it is good to change the closure to a struct and a function and having a type assertion for `CommandReferenceʳ`. 

The others are optional.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:113]: Yep, no worries there.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:137]: This is a canonical way to represent a set in Go: `map[<type of your set>]struct{}`. Then you can use the map to test for presence in set, etc...

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:145]: There are O(100) renderpasses per frame on realistic workloads, so I don't think we need to worry there. If framegraph creation turns out to be slow, we'll measure and optimize accordingly.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:55]: Added a type check.

File: gapis/api/vulkan/framegraph.go
Comment (hevrard) [reply gapis/api/vulkan/framegraph.go:103]: I think such closures are fairly common, but I agree with your comment nonetheless. I've extracted this lambda logic into a new `processSubCommand()` function.

**Review by yalcinmelihyasin — APPROVED** [review]

**Pull request merged**
